File: apps/api/app/schemas/shipment.py
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import (
    AddressValidationStatus,
    NDRStatus,
    PaymentStatus,
    PaymentType,
    RTOStatus,
    ShipmentDelayStatus,
    ShipmentStatus,
    ShopifySyncStatus,
)


class ShipmentCreateRequest(BaseModel):
    order_id: uuid.UUID
    awb: str | None = None
    courier_id: uuid.UUID | None = None
    expected_delivery_date: datetime | None = None


class ShipmentUpdateRequest(BaseModel):
    courier_id: uuid.UUID | None = None
    current_status: ShipmentStatus | None = None
    expected_delivery_date: datetime | None = None
    current_location: str | None = None


class ShipmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    order_id: uuid.UUID
    shiprocket_shipment_id: str | None
    awb: str | None
    courier_id: uuid.UUID | None
    current_status: ShipmentStatus
    delay_status: ShipmentDelayStatus
    ndr_status: NDRStatus | None
    rto_status: RTOStatus | None
    pickup_date: datetime | None
    expected_delivery_date: datetime | None
    actual_delivery_date: datetime | None
    current_location: str | None
    last_tracking_update_at: datetime | None
    source_system: str | None
    shopify_fulfillment_id: str | None
    shopify_sync_status: ShopifySyncStatus
    shopify_sync_error: str | None
    shopify_synced_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ShipmentListResponse(ShipmentResponse):
    """`ShipmentResponse` plus the denormalized fields the Shipments list
    (`GET /shipments`) needs to render order/customer/courier/telecaller
    columns without an N+1 request per row — same pattern as
    `OrderListResponse`/`ShipmentQueueRowResponse`. Only `list_shipments`
    returns this; `GET /shipments/{id}` still returns plain
    `ShipmentResponse`, which already has its own order/customer context
    from the order detail page it's shown alongside.
    """

    order_number: str | None = None
    customer_name: str | None = None
    payment_type: PaymentType | None = None
    confirmed_by_telecaller_name: str | None = None
    courier_name: str | None = None


class ShipmentEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    shipment_id: uuid.UUID
    external_event_id: str | None
    status: str
    location: str | None
    event_timestamp: datetime
    description: str | None
    courier_name: str | None
    source: str
    created_at: datetime


class ShipmentEventCreateRequest(BaseModel):
    external_event_id: str | None = None
    status: str = Field(min_length=1, max_length=100)
    location: str | None = None
    event_timestamp: datetime
    description: str | None = None
    courier_name: str | None = None


class ShipmentQueueRowResponse(BaseModel):
    """One row of the Shipment Queue — confirmed orders awaiting shipment
    processing (`OrderRepository.shipment_queue_query`). Denormalized the
    same way `AssignedOrderResponse`/`OrderListResponse` already are, so
    the table renders without an N+1 request per row. `shipment_*` fields
    are `None` when no shipment has been created for this order yet.
    """

    order_id: uuid.UUID
    order_number: str
    customer_name: str | None
    customer_phone: str | None
    item_summary: str | None
    sku_summary: str | None
    total_quantity: int
    total_amount: Decimal
    payment_type: PaymentType
    payment_status: PaymentStatus
    confirmed_at: datetime | None
    confirmed_by_telecaller_id: uuid.UUID | None
    confirmed_by_telecaller_name: str | None
    shipment_id: uuid.UUID | None = None
    shipment_status: ShipmentStatus | None = None
    awb: str | None = None
    courier_name: str | None = None
    # Outbound Shopify push status for this row's shipment -- lets the
    # queue table offer "Retry Shopify Sync" directly, without opening
    # the shipment. `None` exactly when `shipment_id` is `None`.
    shopify_sync_status: ShopifySyncStatus | None = None
    # The real, numeric Shiprocket order id for this shipment (see
    # `app.services.shiprocket_service.shiprocket_order_id`) -- `None`
    # when the OMS has no reliably-stored one, including whenever
    # `shipment_id` is `None`. Shiprocket has no supported deep-link
    # filter for its "Ready to Ship" page, so the Fulfillment Queue's
    # "Process Shipment"/"Ship Order" actions open that page plain and
    # copy this id to the clipboard, instead of calling any Shiprocket
    # create-shipment API.
    shiprocket_order_id: str | None = None
    # Address validation (Shiprocket-style confidence check) -- see
    # `Order.shipping_address_validation_status` in app/models/order.py.
    # Sourced directly from the SAME `Order` row this whole query already
    # joins in (`OrderRepository.shipment_queue_query`), so showing this
    # on the Fulfillment Queue costs zero extra queries per row -- never
    # `None` when a `shipping_address` exists but hasn't been validated
    # yet; that case is `status=None` (shown as "Validation pending"),
    # not omitted.
    shipping_address_validation_status: AddressValidationStatus | None = None
    shipping_address_validation_score: int | None = None
    shipping_address_validation_reason: str | None = None


class ShipmentSummaryResponse(BaseModel):
    confirmed_awaiting_shipment: int = 0
    total_shipments: int = 0
    pending: int = 0
    picked_up: int = 0
    in_transit: int = 0
    out_for_delivery: int = 0
    delivered: int = 0
    ndr: int = 0
    rto: int = 0
    cancelled: int = 0
    cod: int = 0
    prepaid: int = 0
    todays_shipments: int = 0


class ShipmentStatusBreakdownItem(BaseModel):
    status: str
    count: int


class DailyShipmentTrendPoint(BaseModel):
    date: str
    created: int
    delivered: int


class TelecallerShipmentStats(BaseModel):
    telecaller_id: uuid.UUID
    telecaller_name: str
    confirmed: int
    shipped: int
    delivered: int
    ndr: int = 0
    rto: int = 0


class BulkShipOrdersRequest(BaseModel):
    order_ids: list[uuid.UUID] = Field(min_length=1, max_length=200)


class BulkShipOrderResult(BaseModel):
    order_id: uuid.UUID
    success: bool
    message: str | None
    shipment_id: uuid.UUID | None = None


class BulkShipOrdersResponse(BaseModel):
    shipped_count: int
    failed_count: int
    results: list[BulkShipOrderResult]


class BulkShipValidationResult(BaseModel):
    """One order's dry-run eligibility result from `ShiprocketOperations
    Service.validate_shipment_eligibility` -- never creates anything, just
    classifies. `reason` is `None` exactly when `ready` is `True`.
    """

    order_id: uuid.UUID
    ready: bool
    reason: str | None = None


class ShipmentAnalyticsResponse(BaseModel):
    status_breakdown: list[ShipmentStatusBreakdownItem]
    # % of orders that ever reached CONFIRMED (or any status after it —
    # the state machine is strictly monotonic, so PROCESSING/PACKED/
    # SHIPPED/DELIVERED all imply "was confirmed") AND were never
    # fulfilled directly through Shopify (excluded — those never enter
    # this OMS's own shipment pipeline at all) whose shipment has moved
    # beyond the queue (PICKED_UP or later). See
    # `OrderRepository.confirmation_to_shipment_stats`'s docstring.
    confirmation_to_shipment_rate: float = 0.0
    daily_trend: list[DailyShipmentTrendPoint]
    telecaller_stats: list[TelecallerShipmentStats]

File: apps/web/types/shipment.ts
// Mirrors app/schemas/shipment.py

export type ShipmentStatus =
  | "pending"
  | "picked_up"
  | "in_transit"
  | "out_for_delivery"
  | "delivered"
  | "ndr"
  | "rto_initiated"
  | "rto_delivered"
  | "cancelled"

export type ShipmentDelayStatus = "on_time" | "at_risk" | "delayed" | "unknown"

/** Outbound OMS -> Shopify fulfillment push state (Phase 6) — distinct
 * from `ShipmentStatus` (Shiprocket/logistics) and from an order's own
 * `fulfillment_status` (Shopify's inbound summary).
 */
export type ShopifySyncStatus = "not_applicable" | "pending" | "synced" | "failed"

export interface Shipment {
  id: string
  order_id: string
  shiprocket_shipment_id: string | null
  awb: string | null
  courier_id: string | null
  current_status: ShipmentStatus
  delay_status: ShipmentDelayStatus
  ndr_status: string | null
  rto_status: string | null
  pickup_date: string | null
  expected_delivery_date: string | null
  actual_delivery_date: string | null
  current_location: string | null
  last_tracking_update_at: string | null
  source_system: string | null
  shopify_fulfillment_id: string | null
  shopify_sync_status: ShopifySyncStatus
  shopify_sync_error: string | null
  shopify_synced_at: string | null
  created_at: string
  updated_at: string
  // Present only on rows from `GET /shipments` (`ShipmentListResponse`)
  // -- see `_to_shipment_list_response` in the backend endpoint. Absent
  // (undefined) on a plain `ShipmentResponse`, e.g. `GET /shipments/{id}`.
  order_number?: string | null
  customer_name?: string | null
  payment_type?: string | null
  confirmed_by_telecaller_name?: string | null
  courier_name?: string | null
}

export interface ShipmentEvent {
  id: string
  shipment_id: string
  external_event_id: string | null
  status: string
  location: string | null
  event_timestamp: string
  description: string | null
  courier_name: string | null
  source: string
  created_at: string
}

export interface ShipmentListFilters {
  q?: string
  status?: ShipmentStatus
  courier_id?: string
  date_from?: string
  date_to?: string
}

export const SHIPMENT_STATUS_OPTIONS: { label: string; value: ShipmentStatus }[] = [
  { label: "Pending", value: "pending" },
  { label: "Picked up", value: "picked_up" },
  { label: "In transit", value: "in_transit" },
  { label: "Out for delivery", value: "out_for_delivery" },
  { label: "Delivered", value: "delivered" },
  { label: "NDR", value: "ndr" },
  { label: "RTO initiated", value: "rto_initiated" },
  { label: "RTO delivered", value: "rto_delivered" },
  { label: "Cancelled", value: "cancelled" },
]

/** One row of the Shipment Queue — confirmed orders awaiting shipment
 * processing, a distinct/narrower view from the plain `Shipment` list
 * above (already-processed shipments). `shipment_*` fields are `null`
 * when no shipment has been created for this order yet.
 */
export interface ShipmentQueueRow {
  order_id: string
  order_number: string
  customer_name: string | null
  customer_phone: string | null
  item_summary: string | null
  sku_summary: string | null
  total_quantity: number
  total_amount: string
  payment_type: string
  payment_status: string
  confirmed_at: string | null
  confirmed_by_telecaller_id: string | null
  confirmed_by_telecaller_name: string | null
  shipment_id: string | null
  shipment_status: ShipmentStatus | null
  awb: string | null
  courier_name: string | null
}

export interface ShipmentQueueFilters {
  q?: string
  payment_type?: string
  telecaller_id?: string
  courier_id?: string
  sku?: string
  shipment_status?: ShipmentStatus
  date_from?: string
  date_to?: string
}

export interface ShipmentSummary {
  confirmed_awaiting_shipment: number
  total_shipments: number
  pending: number
  picked_up: number
  in_transit: number
  out_for_delivery: number
  delivered: number
  ndr: number
  rto: number
  cancelled: number
  cod: number
  prepaid: number
  todays_shipments: number
}

export interface ShipmentStatusBreakdownItem {
  status: string
  count: number
}

export interface DailyShipmentTrendPoint {
  date: string
  created: number
  delivered: number
}

export interface TelecallerShipmentStats {
  telecaller_id: string
  telecaller_name: string
  confirmed: number
  shipped: number
  delivered: number
  ndr: number
  rto: number
}

export interface ShipmentAnalytics {
  status_breakdown: ShipmentStatusBreakdownItem[]
  confirmation_to_shipment_rate: number
  daily_trend: DailyShipmentTrendPoint[]
  telecaller_stats: TelecallerShipmentStats[]
}

export interface BulkShipOrderResult {
  order_id: string
  success: boolean
  message: string | null
  shipment_id: string | null
}

export interface BulkShipOrdersResponse {
  shipped_count: number
  failed_count: number
  results: BulkShipOrderResult[]
}
